libbeat/publisher/queue/diskqueue/serialize.go

```diff
@@ -0,0 +1,161 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// Encoding / decoding routines adapted from
+// libbeat/publisher/queue/spool/codec.go.
+
+package diskqueue
+
+import (
+	"bytes"
+	"time"
+
+	"github.com/elastic/beats/v7/libbeat/beat"
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/outputs/codec"
+	"github.com/elastic/beats/v7/libbeat/publisher"
+	"github.com/elastic/go-structform/gotype"
+	"github.com/elastic/go-structform/json"
+)
+
+type eventEncoder struct {
+	buf    bytes.Buffer
+	folder *gotype.Iterator
+}
+
+type eventDecoder struct {
+	buf []byte
+
+	parser   *json.Parser
+	unfolder *gotype.Unfolder
+}
+
+type entry struct {
+	Timestamp int64
+	Flags     uint8
+	Meta      common.MapStr
+	Fields    common.MapStr
+}
+
+func newEventEncoder() *eventEncoder {
+	e := &eventEncoder{}
+	e.reset()
+	return e
+}
+
+func (e *eventEncoder) reset() {
+	e.folder = nil
+
+	visitor := json.NewVisitor(&e.buf)
+	folder, err := gotype.NewIterator(visitor,
+		gotype.Folders(
+			codec.MakeTimestampEncoder(),
+			codec.MakeBCTimestampEncoder(),
+		),
+	)
+	if err != nil {
+		panic(err)
```

Comment:
Same here :)

Reply:
This one is more roundabout than the previous one, but again this error value is a deterministic result of its input value / type, which is a compile-time constant. Error mechanisms are built into `NewUnfolder` / `NewIterator` so we can safely apply arbitrary options dynamically at runtime, but on the "sanitized" input of a fixed list of explicit encoders there's nowhere on the code path that can return non-nil.

I would like if this were more clear -- the logic calls through to the `elastic/go-structform` library and these guarantees aren't currently documented...

These are certainly unrecoverable errors in the abstract. They mean the queue is permanently unable to encode or decode any data, and what's worse they arise from a part of the API that "can't" fail (`queue.Queue.Producer` etc can't return an error). So we are certainly leaning heavily on this being fail-proof based on the compile-time known-good inputs. I don't see much way around this except to write more verbose comments or establish some "nicer" mechanism for triggering a system-wide fatal error, but I'm open to suggestions...

Reply:
I see :) Thank you for explanation.